Bug 1185433: Hide unlisted and disabled add-ons from search results.

kmaglione opened the pull request:
This probably needs a test.

magopian commented on apps/addons/models.py:
```diff
@@ -383,6 +383,20 @@ class Addon(amo.models.OnChangeMixin, amo.models.ModelBase):
     with_unlisted = AddonManager(include_unlisted=True)
     objects = AddonManager()
 
+    # Like the above, but for ElasticSearch, rather than Django Managers.
```
"Like the above"? Like what above? `# "Managers" for ElasticSearch.`?

kmaglione replied:
The above three managers, `with_unlisted`, `unfiltered`, and `objects`

magopian commented on apps/addons/cron.py:
```diff
@@ -453,9 +453,7 @@ def reindex_addons(index=None, addon_type=None):
            .filter(_current_version__isnull=False,
-                   status__in=amo.VALID_STATUSES,
-                   is_listed=True,
-                   disabled_by_user=False))
```
This was there before the commit you're talking about in the bug (https://github.com/mozilla/olympia/commit/ae2e2cb). I guess it was also faulty?

kmaglione replied:
I don't think it was faulty before. It wasn't really necessary to index these previously. But now we have bad data in the index for these add-ons, so we need to do at least one run that doesn't exclude them.

magopian commented on apps/addons/search.py:
```diff
@@ -28,8 +27,8 @@ def extract(addon):
-             'hotness', 'is_disabled', 'premium_type')
-    d = dict(zip(attrs, attrgetter(*attrs)(addon)))
+             'hotness', 'is_disabled', 'is_listed', 'premium_type')
+    d = {attr: getattr(addon, attr) for attr in attrs}
```
Much better :+1:

magopian commented on apps/search/views.py:
```diff
@@ -165,16 +164,14 @@ def queryset(self):
             elif len(q) > 2:
                 # Oh, how I wish I could elastically exclude terms.
                 # (You can now, but I forgot why I was complaining to
                 # begin with.)
```
Is this weird comment still useful?

kmaglione replied:
Heh. I don't know. I was wondering that too, and considered removing it, but I wasn't quite sure what it meant.

But since you also seem not to know what it means, I guess I'll remove it.

magopian commented:
This looks much better! Thanks!

Needs to fix the current failing tests, and yeah, maybe add one (or more) tests.

r+wc

kmaglione commented:
OK, I added a couple of tests, and renamed the search methods to `search_public()`, `search_with_unlisted()` and `.search()` to avoid having to update dozens of tests which aren't actually related to add-on search at all. I had to change the manager we passed to `index_objects` so it would update unlisted add-ons.

I also added a migration to reindex the add-ons in ES.

Can you have another look?

magopian commented on apps/addons/models.py:
```diff
@@ -383,6 +383,16 @@ class Addon(amo.models.OnChangeMixin, amo.models.ModelBase):
+    def search_with_unlisted(cls):
+        return cls.search().filter(
+            is_disabled=False, status__in=amo.REVIEWED_STATUSES)
+
```
I didn't realize before, but we usually put any method below the `Meta` class.

magopian commented on apps/addons/cron.py:
```diff
@@ -453,9 +453,7 @@ def reindex_addons(index=None, addon_type=None):
     from . import tasks
     ids = (Addon.objects.values_list('id', flat=True)
```
Doesn't this need to be `Addon.with_unlisted.values_list...`?

magopian replied:
Actually, it doesn't. Sorry about that.

magopian replied:
Or maybe it does? I'm lost now. You say it needs at least one run that doesn't exclude them, but then you need to include them, right? Maybe you even need `Addon.unfiltered....`?

kmaglione replied:
You're right, it does. With `objects`, it fixes the `disabled_by_user` cases but not the unlisted cases. Nice catch.

Don't think we need unfiltered, since the hook for deleted add-ons never stopped working.

magopian commented on apps/search/views.py:
```diff
@@ -443,9 +437,8 @@ def search(request, tag_name=None, template=None):
-          .filter(status__in=amo.REVIEWED_STATUSES, is_disabled=False,
-                  app=APP.id)
+    qs = (Addon.search_public()
+          .filter(app=APP.id)
```
Fits on a single line

magopian commented on apps/addons/tests/test_models.py:
```diff
@@ -2338,35 +2338,66 @@ def cleanup(self):
         self.refresh()
-        eq_(Addon.search().count(), 1)
-        eq_(Addon.search().query(name='woo')[0].id, addon.id)
+        eq_(Addon.search_public().count(), 1)
```
You can use simple `assert`s in olympia, we're using pytest ;)

kmaglione replied:
Eh, I was trying to make minimal changes, but OK.

magopian commented on apps/addons/tests/test_models.py:
```diff
@@ -2338,35 +2338,66 @@ def cleanup(self):
+        addon.save()
+        self.refresh()
+
+        eq_(Addon.search_public().count(), 0)
```
May I ask for a test for add-ons submitted as unlisted? (created as unlisted)

kmaglione replied:
Yeah, makes sense. I was assuming we already had one, but it looks like we don't.

kmaglione replied:
Added a test that add-ons created as unlisted do not show up, but do after switching to listed (even though we don't support that yet).

magopian replied:
I don't see it? Did you push?

kmaglione replied:
On Tue, Jul 21, 2015 at 12:20:57PM -0700, Mathieu Agopian wrote:
>I don't see it? Did you push?

Yeah, but about a minute after I wrote the comment.

magopian commented:
A few remaining nitpicks ;)

Thanks!

r+wc

magopian commented:
Excellent! As soon as travis is finished (and green), r+
